### PR TITLE
Implement AOS-47 & AOS-51

### PR DESCRIPTION
**Pull request by xxchan**
part of https://github.com/aos-dev/go-storage/issues/558

### CODE REVIEW

File: storage.go [hunk 1]
```diff
@@ -135,7 +136,15 @@ func (s *Storage) fetch(ctx context.Context, path string, url string, opt pairSt
 	defer resp.Body.Close()
 
 	if resp.StatusCode != http.StatusOK {
-		return fmt.Errorf("fetch from url %s expected %d, but got %d", url, http.StatusOK, resp.StatusCode)
+		switch resp.StatusCode {
```
Comment (Xuanwo) [inline storage.go:139]: We don't know the real meaning of those HTTP status codes, so don't need to handle them, just return ErrUnexpected for all of them.

File: utils.go [hunk 2]
```diff
@@ -80,14 +81,23 @@ func newStorager(pairs ...typ.Pair) (store *Storage, err error) {
 }
 
 func formatError(err error) error {
+	if _, ok := err.(services.AosError); ok {
+		return err
+	}
+	if errors.Unwrap(err) != nil {
```
Comment (Xuanwo) [inline utils.go:87]: Why do we need to check this?

**Comment by xxchan** [comment]
update corresponding to discussion
https://matrix.to/#/!l8rnxDAEOzBxO84I:aos.dev/$Yn0L0tkor1IEBUOYmWQFwwosn59V58Svm2ArKtHDKCk?via=aos.dev&via=matrix.org&via=t2bot.io

File: storage.go [hunk 3]
```diff
@@ -135,7 +135,15 @@ func (s *Storage) fetch(ctx context.Context, path string, url string, opt pairSt
...
+		case 403:
+			err = os.ErrPermission
+		case 404:
+			err = os.ErrNotExist
+		default:
+			return fmt.Errorf("fetch from url %s expected %d, but got %d", url, http.StatusOK, resp.StatusCode)
+		}
+		return fmt.Errorf("%w: fetch from url %s expected %d, but got %d", err, url, http.StatusOK, resp.StatusCode)
```
Comment (Xuanwo) [inline storage.go:146]: Should be removed.

Comment (Xuanwo) [reply storage.go:146]: I thought we don't need to wrap the error here?

Comment (xxchan) [reply storage.go:146]: I think wrap it with the same message as "the unknown case" may be more consistent?

Comment (xxchan) [reply storage.go:146]: e.g., 
code 500: error message is "fetch from url..."
but code 404: error message is "not found"

Comment (Xuanwo) [reply storage.go:146]: OK, Accepted.

**Review by Xuanwo — APPROVED** [review]

**Pull request merged**